embassy-stm32/src/rcc/f2.rs

```diff
@@ -226,30 +406,65 @@ impl Default for Config {
     }
 }
 
-#[inline]
-unsafe fn enable_hse(source: HSESrc) {
-    RCC.cr().write(|w| {
-        w.set_hsebyp(match source {
-            HSESrc::Bypass => true,
-            HSESrc::Crystal => false,
+pub(crate) unsafe fn init(config: Config) {
+    // Make sure HSI is enabled
+    RCC.cr().write(|w| w.set_hsion(true));
+    while !RCC.cr().read().hsirdy() {}
+
+    if let Some(hse_config) = config.hse {
+        RCC.cr().modify(|w| {
+            w.set_hsebyp(match hse_config.source {
+                HSESrc::Bypass => true,
+                HSESrc::Crystal => false,
+            });
+            w.set_hseon(true)
         });
-        w.set_hseon(true)
+        while !RCC.cr().read().hserdy() {}
+    }
+
+    let pll_src_freq = match config.pll_mux {
+        PLLSrc::HSE => {
+            config
+                .hse
+                .expect("HSE must be configured to be used as PLL input")
+                .frequency
+        }
+        PLLSrc::HSI => HSI,
+    };
+
+    // Reference: STM32F215xx/217xx datasheet Table 33. Main PLL characteristics
+    let pll_clocks = config.pll.clocks(pll_src_freq);
+    assert!(Hertz(950_000) <= pll_clocks.in_freq && pll_clocks.in_freq <= Hertz(2_100_000));
+    assert!(Hertz(192_000_000) <= pll_clocks.vco_freq && pll_clocks.vco_freq <= Hertz(432_000_000));
+    assert!(
+        Hertz(24_000_000) <= pll_clocks.main_freq && pll_clocks.main_freq <= Hertz(120_000_000)
+    );
+    // USB actually requires == 48 MHz, but other PLL48 peripherals are fine with <= 48MHz
+    assert!(pll_clocks.pll48_freq <= Hertz(48_000_000));
+
+    RCC.pllcfgr().write(|w| {
+        w.set_pllsrc(config.pll_mux.into());
+        w.set_pllm(config.pll.pre_div.0);
+        w.set_plln(config.pll.mul.0);
+        w.set_pllp(config.pll.main_div.into());
+        w.set_pllq(config.pll.pll48_div.0);
     });
-    while !RCC.cr().read().hserdy() {}
-}
 
-pub(crate) unsafe fn init(config: Config) {
     let (sys_clk, sw) = match config.mux {
         ClockSrc::HSI => {
-            // Enable HSI
-            RCC.cr().write(|w| w.set_hsion(true));
-            while !RCC.cr().read().hsirdy() {}
-
+            assert!(config.hsi, "HSI must be enabled to be used as system clock");
             (HSI, Sw::HSI)
         }
-        ClockSrc::HSE(freq, source) => {
-            enable_hse(source);
-            (freq, Sw::HSE)
+        ClockSrc::HSE => {
+            let hse_config = config
+                .hse
+                .expect("HSE must be configured to be used as system clock");
```

Comment:
could you use `unwrap!`, `panic!` or `assert!` or similar instead of `.expect()`? `expect()` stores the error message as a string, while the others can use defmt's string tokenization.

Reply:
Sure! I naively thought assert did that too so I didn't give the error handling style much thought...I'll update the example too since it uses `.unwrap()` and not `unwrap!`